orchagent/fdborch.cpp

```diff
@@ -451,23 +502,54 @@ void FdbOrch::flushFDBEntries(sai_object_id_t bridge_port_oid,
     }
 }
 
+void FdbOrch::notifyObserversFDBFlush(Port &port, sai_object_id_t& bvid)
+
+{
+    FdbFlushUpdate flushUpdate;
+    flushUpdate.port = port;
+
+    for (auto itr = m_entries.begin(); itr != m_entries.end(); ++itr)
+    {
+        if ((itr->port_name == port.m_alias) && 
+            (itr->bv_id == bvid))
+        {
+            SWSS_LOG_INFO("Adding MAC learnt on [ port:%s , bvid:0x%" PRIx64 "]\
+                           to ARP flush", port.m_alias.c_str(), bvid);
+            FdbEntry entry;
+            entry.mac = itr->mac;
+            entry.bv_id = itr->bv_id;
+            flushUpdate.entries.push_back(entry);
+        }
+    }
+
+    if (!flushUpdate.entries.empty())
+    {
+        for (auto observer: m_observers)
+        {
+            observer->update(SUBJECT_TYPE_FDB_FLUSH_CHANGE, &flushUpdate);
+        }
+    }
+}
+
 void FdbOrch::updateVlanMember(const VlanMemberUpdate& update)
 {
     SWSS_LOG_ENTER();
 
     if (!update.add)
     {
+        SWSS_LOG_ERROR("Vasant: updateVlanMember");
         swss::Port vlan = update.vlan;
         swss::Port port = update.member;
         flushFDBEntries(port.m_bridge_port_id, vlan.m_vlan_info.vlan_oid);
+        notifyObserversFDBFlush(port, vlan.m_vlan_info.vlan_oid);
```

Comment:
I think we should not need notifyObserversFDBFlush(),  but do it in flushFDBEntries() as you defined.   Maybe rename flushFDBEntries to flushFDBAndNeighbor.   We should change the order in portorch so that we do notify first then then update the membership etc.

Reply:
notifyObserversFDBFush() means whoever is observind FDB orch, notify them that FDB flush happened. This is in line with the SUBJECT types of observer pattern. For example SUBJECT_TYPE_FDB_FLUSH_CHANGE. Now, how observers react to this notification is left to the observer. In this case, neighbor orch is an observer and he reacts by flushing arp entries. But that is neighbororch's responsibility, we shouldn't bring it here. Sorry I dint get what you mean by "notify first and then update the membership"

Reply:
My point is in the portsorch, we have "flushFDBEntries" which does the ARP flush, If we want to follow that, I recommend moving the notifyObserversFDBFlush into flushFDBEntries at fdborch too.

Reply:
As you know, implementing flushFDBEntries in portsorch was NOT correct, hence I am moving that to fdbOrch in the upstream PR. Once its done flushFDBEntries() method in fdborch will be used for two cases 
1. Flush FDB entries with valid bridge_port_id and valid bvid. (this happens on receiving vlanMemberUpdate)
2. Flush FDB entries with valid bridge_port_id only (this happens on receiving portStateChangeUpdate)

Today notifyObserversFDBFlush() applies only for case 1, but after that change I will update this function to handle the case when bvid == SAI_NULL_OBJECT_ID. And notifyObserversFDBFlush will be called from the function updatePortOperState() function. Now if you notice, updateVlanMember() and updatePortOperState() both will call flushFDBEntries() and followed by notifyObserversFDBFlush(). Moving forward, if you do NOT want flush ARP in case of vlanMemberUpdate(), you can just delete that call. And if you are calling flushFDBEntries() some other place where you do NOT want to flush ARP, you can do so without handling special case flushFDbEntries(). Again, I think this is modular, instead of nesting things. Please suggest if you have other idea, or see any flaw in this.

Reply:
As talked, it is fine for now, when the function settled, it would be better to clean things up.